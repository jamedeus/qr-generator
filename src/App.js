import React, { useState, useEffect } from 'react';
import ContactForm from './ContactForm';
import WifiForm from './WifiForm';
import LinkForm from './LinkForm';


function App() {
    // Default to contact QR code form
    const [qrType, setQrType] = useState('contact');

    // State object for base64 QR code string
    const [qrString, setQrString] = useState('');

    // State object for QR code visibility (controls fade effect)
    const [qrVisible, setQrVisible] = useState(false);

    // Takes type string, shows correct form and hides old QR code (if present)
    const showForm = (type) => {
        setQrType(type);
        resetQR();
    }

    // Fades out QR code, clears QR string once animation complete
    const resetQR = () => {
        setQrVisible(false);
        setTimeout(() => {
            setQrString("");
        }, 468);
    }

    // Add fade in/out classes when QR code visibility changes
    // Mobile: Scroll to bottom when QR shown, scroll to top when hidden
    useEffect(() => {
        const outputColumn = document.getElementById('output_column');
        const downloadButton = document.getElementById('download');
        if (qrVisible) {
            outputColumn.classList.remove('fade-out');
            outputColumn.classList.add('fade-in');
            downloadButton.scrollIntoView({behavior: "smooth"});

        } else {
            outputColumn.classList.remove('fade-in');
            outputColumn.classList.add('fade-out');
            window.scroll({ top: 0, left: 0, behavior: 'smooth' });
        }
    });

    // Called by generate buttons, takes submit event as arg
    async function generate(e) {
        e.preventDefault();

        // Get form data, add QR type
        const data = Object.fromEntries(new FormData(e.target).entries())
        data['type'] = `${qrType}-qr`;
        console.log(data);

        // Post form data, receive base64 PNG
        var result = await fetch('/generate', {
            method: 'POST',
            body: JSON.stringify(data),
            headers: {
                'Accept': 'application/json, text/plain, */*',
                'Content-Type': 'application/json'
            }
        });

        const outputColumn = document.getElementById('output_column');
        outputColumn.classList.remove('d-none');
        outputColumn.classList.add('d-flex');

        // Write result string to state object (shows QR image)
        result = await result.text();
        setQrString(result);
        setQrVisible(true);
        console.log(result);
    }

    // Decodes base64 image string from event to binary, serves download
    function downloadQR(event) {
        // Decode base64 image data to binary
        const imageData = atob(qrString);

        // Create Uint8Array with same length
        const imageBuffer = new ArrayBuffer(imageData.length);
        const imageBytes = new Uint8Array(imageBuffer);

        // Read bytes into array
        for (let i = 0; i < imageData.length; i++) {
            imageBytes[i] = imageData.charCodeAt(i);
        }

        // Create blob object from buffer, set MIME type
        const qr = new Blob([imageBytes], { type: 'image/png' });

        // Set download url and filename, downloads immediately
        event.target.href = URL.createObjectURL(qr);
        event.target.download = `${qrType}-qr.png`;
    }

    return (
        <div className="d-flex flex-column vh-100">
            {/* Fixed nav bar reserves space */}
            <nav className="navbar navbar-dark bg-primary fixed-top">
                <div className="container-fluid">
                    <button type="button" className="btn" style={{visibility: "hidden"}}><i className="bi-list"></i></button>
                    <h1 className="mb-0 navbar-brand mx-auto">QR Code Generator</h1>

                    {/* Dropdown to select QR Code type */}
                    <div className="dropdown my-auto">
                        <button type="button" className="btn" id="menu_button" data-bs-toggle="dropdown" aria-expanded="false"><i className="bi-list" style={{color: "#fff"}}></i></button>
                        <ul id="settings_menu" className="dropdown-menu dropdown-menu-end" aria-labelledby="settings_button">
                            <li>
                                <a className={"dropdown-item " + (qrType === "contact" ? "active" : "")} onClick={(() => {showForm('contact')})}>
                                    <i className="bi bi-person-lines-fill me-3"></i>Contact
                                </a>
                            </li>
                            <li>
                                <a className={"dropdown-item " + (qrType === "wifi" ? "active" : "")} onClick={(() => {showForm('wifi')})}>
                                    <i className="bi bi-wifi me-3"></i>Wifi
                                </a>
                            </li>
                            <li>
                                <a className={"dropdown-item " + (qrType === "link" ? "active" : "")} onClick={(() => {showForm('link')})}>
                                    <i className="bi bi-link-45deg me-3"></i>Link
                                </a>
                            </li>
                        </ul>
                    </div>
                </div>
            </nav>

            {/* Container reserves remaining space */}
            <div className="container d-flex flex-column h-100">
                {/* Side by side on desktop, stacked on mobile */}
                <div className="row h-100">
                    {/* Same height as container to vertically center contents */}
                    <div className="col-md-6 d-flex flex-column justify-content-center py-3 h-100">
                        {(() => {
                            switch(qrType) {
                                case "contact":
                                    return <ContactForm onSubmit={generate} />;
                                case "wifi":
                                    return <WifiForm onSubmit={generate} />;
                                case "link":
                                    return <LinkForm onSubmit={generate} />;
                                default:
                                    return null;
                            }
                        })()}
                    </div>

                    <div id="output_column" className="col-md-6 d-none flex-column justify-content-center align-items-center py-3 h-100">
                        {/* Keep QR vertically centered, hidden button negates download button impact on layout */}
                        <a style={{visibility: "hidden"}} className="btn btn-primary mb-3">Download</a>

                        {/* Output image + download button */}
                        <img id="output" src={"data:image/png;base64," + qrString}></img>
                        <a id="download" className="btn btn-primary mb-3" onClick={downloadQR}>Download</a>
                    </div>
                </div>
            </div>
        </div>
    )
}


export default App;
