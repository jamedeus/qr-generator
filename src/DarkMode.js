import React, { createContext, useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import Button from 'react-bootstrap/Button';
import { BrightnessHighFill, MoonFill } from 'react-bootstrap-icons';

const ThemeContext = createContext();

export const ThemeProvider = ({ children }) => {
    const [theme, setTheme] = useState('light');

    // Setup, runs once when mounted
    useEffect(() => {
        // Check system theme once, override default (light) if user prefers dark
        if (window.matchMedia && window.matchMedia('(prefers-color-scheme: dark)').matches) {
            setTheme('dark');
        }

        // Listen for system theme changes, switch to dark/light mode responsively
        window.matchMedia('(prefers-color-scheme: dark)').addListener((e) => {
            if (e.matches) { // Returns True for dark mode, False otherwise
                setTheme('dark');
            } else {
                setTheme('light');
            }
        });
    }, []);

    // Change theme data attribute when state object changes
    useEffect(() => {
        document.documentElement.setAttribute('data-bs-theme', theme);
    }, [theme]);

    return (
        <ThemeContext.Provider value={{ theme, setTheme }}>
            {children}
        </ThemeContext.Provider>
    );
};

ThemeProvider.propTypes = {
    children: PropTypes.node,
};

export const DarkModeButton = () => {
    // Get theme state + hook to change
    const { theme, setTheme } = useContext(ThemeContext);

    switch(theme) {
        case "light":
            return (
                <Button
                    className="my-auto"
                    title="Switch to dark mode"
                    onClick={(() => {setTheme('dark');})}
                >
                    <MoonFill />
                </Button>
            );
        case "dark":
            return (
                <Button
                    className="my-auto"
                    title="Switch to light mode"
                    onClick={(() => {setTheme('light');})}
                >
                    <BrightnessHighFill />
                </Button>
            );
    }
};
