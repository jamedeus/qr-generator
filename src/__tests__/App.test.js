import renderer from 'react-test-renderer';
import userEvent from "@testing-library/user-event";
import { act } from 'react-dom/test-utils';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { ThemeProvider } from '../DarkMode.js';
import App from '../App';

describe('App', () => {
    let app, nav, form, submitButton;
    let getByText, getByPlaceholderText, queryByPlaceholderText;

    // Setup: render App component, save references to buttons etc
    beforeEach(() => {
        // Mock window.scroll
        window.scroll = jest.fn();

        // Mock fetch function
        global.fetch = jest.fn(() =>
            Promise.resolve({
                ok: true,
                text: () => Promise.resolve('mock_image_string')
            })
        );

        // Render component
        app = render(
            <ThemeProvider>
                <App />
            </ThemeProvider>
        );

        // Export references to each component used by tests
        nav = app.getByText(/QR Code Generator/).parentElement;
        submitButton = app.getByText(/generate/i);
        form = submitButton.parentElement.parentElement;

        // Export query functions
        getByText = app.getByText;
        getByPlaceholderText = app.getByPlaceholderText;
        queryByPlaceholderText = app.queryByPlaceholderText;
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('matches snapshot', () => {
        const component = renderer.create(
            <ThemeProvider>
                <App />
            </ThemeProvider>
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders the correct form when dropdown buttons are clicked', () => {
        // Click top-right menu button (open dropdown, render options inside)
        act(() => {
            nav.childNodes[2].childNodes[0].click();
        });

        // Get reference to each button that just rendered
        const ContactButton = nav.childNodes[2].childNodes[1].childNodes[0];
        const WifiButton = nav.childNodes[2].childNodes[1].childNodes[1];
        const LinkButton = nav.childNodes[2].childNodes[1].childNodes[2];

        // Should show contact by default
        expect(queryByPlaceholderText('First Name').nodeName).toBe('INPUT');
        expect(queryByPlaceholderText('SSID')).toBeNull();
        expect(queryByPlaceholderText('URL')).toBeNull();

        // Click wifi button, should show wifi form + hide other forms
        act(() => {
            fireEvent.click(WifiButton);
        });
        expect(queryByPlaceholderText('First Name')).toBeNull();
        expect(queryByPlaceholderText('SSID').nodeName).toBe('INPUT');
        expect(queryByPlaceholderText('URL')).toBeNull();

        // Click link button, should show link form + hide other forms
        act(() => {
            fireEvent.click(LinkButton);
        });
        expect(queryByPlaceholderText('First Name')).toBeNull();
        expect(queryByPlaceholderText('SSID')).toBeNull();
        expect(queryByPlaceholderText('URL').nodeName).toBe('INPUT');

        // Click contact button, should show contact form + hide other forms
        act(() => {
            fireEvent.click(ContactButton);
        });
        expect(queryByPlaceholderText('First Name').nodeName).toBe('INPUT');
        expect(queryByPlaceholderText('SSID')).toBeNull();
        expect(queryByPlaceholderText('URL')).toBeNull();

    });

    it('sends correct request when valid form is submitted', async () => {
        // Confirm form not validated, output column not rendered
        expect(form.classList).not.toContainEqual('was-validated');
        expect(app.container.querySelector('img')).toBeNull();

        // Populate all fields, click generate button
        getByPlaceholderText('First Name').value = 'first';
        getByPlaceholderText('Last Name').value = 'last';
        getByPlaceholderText('Email').value = 'first.last@mail.com';
        getByPlaceholderText('Phone').value = '(123) 456-7890';
        await userEvent.click(submitButton);

        // Confirm correct data posted to /generate endpoint
        expect(global.fetch).toHaveBeenCalledWith('/generate', expect.objectContaining({
            method: 'POST',
            body: JSON.stringify({
                "firstName": "first",
                "lastName": "last",
                "email": "first.last@mail.com",
                "phone": "(123) 456-7890",
                "type": "contact-qr"
            }),
            headers: {
                'Accept': 'application/json, text/plain, */*',
                'Content-Type': 'application/json'
            }
        }));

        // Confirm form is now validated, output column rendered, img has correct source
        expect(form.classList).toContainEqual('was-validated');
        expect(app.container.querySelector('img')).not.toBeNull();
        expect(app.container.querySelector('img').src).toBe(
            'data:image/png;base64,mock_image_string'
        );
    });

    it('does not make a request when form is invalid', async () => {
        // Confirm form not validated, output column not rendered
        expect(form.tagName).toBe('FORM');
        expect(form.classList).not.toContainEqual('was-validated');
        expect(app.container.querySelector('img')).toBeNull();

        // Click generate button without filling in fields
        act(() => {
            fireEvent.click(submitButton);
        });

        // Confirm fetch was NOT called
        expect(global.fetch).not.toHaveBeenCalled();

        // Confirm form is now validated, but output column still did not render
        expect(form.classList).toContainEqual('was-validated');
        expect(app.container.querySelector('img')).toBeNull();
    });

    it('shows an alert if an error is received from backend', async () => {
        // Spy on window.alert, mock fetch function to return expected error
        const alertSpy = jest.spyOn(window, 'alert');
        global.fetch = jest.fn(() =>
            Promise.resolve({
                ok: false,
                status: 400,
                text: () => Promise.resolve('Unsupported QR code type')
            })
        );

        // Populate all fields, click generate button
        getByPlaceholderText('First Name').value = 'first';
        getByPlaceholderText('Last Name').value = 'last';
        getByPlaceholderText('Email').value = 'first.last@mail.com';
        getByPlaceholderText('Phone').value = '(123) 456-7890';
        await userEvent.click(submitButton);

        // Confirm window.alert was called, output column did not render
        expect(alertSpy).toHaveBeenCalled();
        expect(app.container.querySelector('img')).toBeNull();
    });

    it('correctly handles downloadQR function', async () => {
        // Mock atob to ignore input and return a known value
        jest.spyOn(window, 'atob').mockImplementation(() => 'base64_string');

        // Mock Blob function, save reference to confirm data
        const blobMock = jest.fn();
        global.Blob = blobMock;

        // Mock createObjectURL to return a known value
        URL.createObjectURL = jest.fn(() => 'blob:url');

        // Populate fields and click generate to render download button
        getByPlaceholderText('First Name').value = 'first';
        getByPlaceholderText('Last Name').value = 'last';
        getByPlaceholderText('Email').value = 'first.last@mail.com';
        getByPlaceholderText('Phone').value = '(123) 456-7890';
        await userEvent.click(submitButton);

        // Click download button
        const downloadButton = app.getByText(/download/i);
        act(() => {
            fireEvent.click(downloadButton);
        });

        // Confirm Blob was created with correct data
        expect(blobMock).toMatchSnapshot();

        // Confirm correct URL was generated
        expect(URL.createObjectURL).toHaveBeenCalled();
        expect(downloadButton.href).toBe('blob:url');

        // Confirm correct filename
        expect(downloadButton.download).toBe('contact-qr.png');
    });

    it('unmounts QR code image when form type is changed', async () => {
        // Populate all fields, click generate button
        getByPlaceholderText('First Name').value = 'first';
        getByPlaceholderText('Last Name').value = 'last';
        getByPlaceholderText('Email').value = 'first.last@mail.com';
        getByPlaceholderText('Phone').value = '(123) 456-7890';
        await userEvent.click(submitButton);

        // Confirm output column was rendered, img has expected source
        expect(form.classList).toContainEqual('was-validated');
        expect(app.container.querySelector('img')).not.toBeNull();
        expect(app.container.querySelector('img').src).toBe(
            'data:image/png;base64,mock_image_string'
        );

        // Click top-right menu button (open dropdown, render options inside)
        act(() => {
            nav.childNodes[2].childNodes[0].click();
        });

        // Switch to wifi form
        act(() => {
            const WifiButton = nav.childNodes[2].childNodes[1].childNodes[1];
            fireEvent.click(WifiButton);
        });

        // Confirm fade-exit class is added to output column
        await waitFor(() => {
            expect(
                app.container.querySelector('img').parentElement.classList
            ).toContain('fade-exit');
        }, { timeout: 50});


        // Confirm output column is umounted after animation completes
        await waitFor(() => {
            expect(app.container.querySelector('img')).toBeNull();
        }, { timeout: 500 });

        // Confirm new form is not validated
        const newForm = getByText(/generate/i).parentElement.parentElement;
        expect(newForm.classList).not.toContainEqual('was-validated');
    });
});
