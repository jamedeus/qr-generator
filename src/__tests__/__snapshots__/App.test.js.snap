// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App correctly handles downloadQR function 1`] = `
[MockFunction] {
  "calls": [
    [
      [
        Uint8Array [
          98,
          97,
          115,
          101,
          54,
          52,
          95,
          115,
          116,
          114,
          105,
          110,
          103,
        ],
      ],
      {
        "type": "image/png",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`App matches snapshot 1`] = `
<div
  className="d-flex flex-column vh-100"
>
  <nav
    className="bg-primary navbar navbar-expand navbar-dark fixed-top"
  >
    <div
      className="container-fluid"
    >
      <button
        className="my-auto btn btn-primary"
        disabled={false}
        onClick={[Function]}
        title="Switch to dark mode"
        type="button"
      >
        <svg
          className="mb-1"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 .278a.768.768 0 0 1 .08.858 7.208 7.208 0 0 0-.878 3.46c0 4.021 3.278 7.277 7.318 7.277.527 0 1.04-.055 1.533-.16a.787.787 0 0 1 .81.316.733.733 0 0 1-.031.893A8.349 8.349 0 0 1 8.344 16C3.734 16 0 12.286 0 7.71 0 4.266 2.114 1.312 5.124.06A.752.752 0 0 1 6 .278z"
          />
        </svg>
      </button>
      <span
        className="mx-auto navbar-brand"
      >
        QR Code Generator
      </span>
      <div
        className="dropdown"
      >
        <button
          aria-expanded={false}
          className="my-auto dropdown-toggle btn btn-primary"
          disabled={false}
          id="react-aria-:r0:"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="mb-1"
            fill="currentColor"
            height="1em"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.5 12a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 0 1H3a.5.5 0 0 1-.5-.5zm0-4a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 0 1H3a.5.5 0 0 1-.5-.5zm0-4a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 0 1H3a.5.5 0 0 1-.5-.5z"
              fillRule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
  </nav>
  <div
    className="d-flex flex-column h-100 container"
  >
    <div
      className="h-100 row"
    >
      <div
        className="d-flex flex-column justify-content-center py-3 h-100 col-md-6"
      >
        <form
          className=""
          noValidate={true}
          onSubmit={[Function]}
        >
          <div
            className="mb-3 form-floating"
          >
            <input
              className="form-control"
              name="firstName"
              placeholder="First Name"
              required={true}
              type="text"
            />
            <label>
              First Name
            </label>
          </div>
          <div
            className="mb-3 form-floating"
          >
            <input
              className="form-control"
              name="lastName"
              placeholder="Last Name"
              required={true}
              type="text"
            />
            <label>
              Last Name
            </label>
          </div>
          <div
            className="mb-3 form-floating"
          >
            <input
              className="form-control"
              name="email"
              onKeyDown={[Function]}
              placeholder="Email"
              required={true}
              type="email"
            />
            <label>
              Email
            </label>
          </div>
          <div
            className="mb-3 form-floating"
          >
            <input
              className="form-control"
              name="phone"
              onChange={[Function]}
              pattern="^(\\d{10}|\\(\\d{3}\\) \\d{3}-\\d{4})$"
              placeholder="Phone"
              required={true}
              type="tel"
            />
            <label>
              Phone
            </label>
          </div>
          <div
            className="d-flex"
          >
            <button
              className="mx-auto btn btn-primary"
              disabled={false}
              type="submit"
            >
              Generate
            </button>
          </div>
        </form>
      </div>
      <div
        className="d-none flex-column justify-content-center py-3 h-100 col-md-6"
      >
        <a
          className="my-3 invisible btn btn-primary"
          disabled={false}
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          D
        </a>
        <img
          src="data:image/png;base64,"
        />
        <a
          className="my-3 mx-auto btn btn-primary"
          disabled={false}
          href="#"
          onClick={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Download
        </a>
      </div>
    </div>
  </div>
</div>
`;
